test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java

```diff
@@ -116,16 +129,24 @@ public TestCase nodes(int nodeCount)
             return this;
         }
 
-        public TestCase upgrade(Major initial, Major ... upgrade)
+        public TestCase upgrade(Major from)
         {
-            this.upgrade.add(new TestVersions(versions.getLatest(initial),
-                                              Arrays.stream(upgrade)
-                                                    .map(versions::getLatest)
-                                                    .toArray(Version[]::new)));
+            return upgrade(from, null);
+        }
+
+        public TestCase upgrade(Major from, @Nullable Major to)
+        {
+            SUPPORTED_UPGRADE_PATHS.stream()
+                .filter(upgradePath -> (upgradePath.left.compareTo(from) >= 0 && (null == to || upgradePath.right.compareTo(to) <= 0)))
```

Comment:
If I understand this code correctly, we might want to have strict comparisons here: `upgradePath.left == to`. I'm assuming we're looking for the exact matches, otherwise by adding upgrade from one lower version to another we inadvertently allow other upgrades.

Reply:
The idea here is to add additional upgrades.

For example, if I define the test to be valid for upgrades from `v30` and to `v4X` that means we should be testing the upgrade paths:
 - `v30` -> `v3X`
 - `v30` -> `v40`
 - `v3X` -> `v40`
 - `v40` -> `v4X`

Reply:
I've renames both methods to `upgradesFrom(from)` and `upgrades(from,to)`
with the intentional that the plural verb helps indicate the change and difference.